csi/src/controller.rs

```diff
@@ -723,333 +674,335 @@ impl Controller for ControllerImpl {
+                        return Ok(resp);
+                    } else {
+                        debug!("no snapshot name={} found", snap_name);
+                    }
+                }
+                Err((rpc_status_code, e)) => {
+                    debug!(
+                        "failed to find existing snapshot name={}, the error is: {}",
                         snap_name,
+                        util::format_anyhow_error(&e),
                     );
-                    return util::success(&ctx, sink, resp);
-                } else {
-                    debug!("no snapshot name={} found", snap_name);
+                    return Err((rpc_status_code, e));
                 }
             }
-            Err((rpc_status_code, e)) => {
-                debug!(
-                    "failed to find existing snapshot name={}, the error is: {}",
-                    snap_name,
-                    util::format_anyhow_error(&e),
-                );
-                return util::fail(&ctx, sink, rpc_status_code, &e);
-            }
-        }
 
-        match self.meta_data.get_volume_by_id(src_vol_id) {
-            Some(src_vol) => {
-                let node_res = self.meta_data.get_node_by_id(&src_vol.node_id);
-                if let Some(node) = node_res {
-                    let client = MetaData::build_worker_client(&node);
-                    let create_res = client.worker_create_snapshot(&req).context(format!(
-                        "failed to create snapshot name={} on node ID={}",
-                        snap_name, src_vol.node_id,
-                    ));
-                    match create_res {
-                        Ok(r) => util::success(&ctx, sink, r),
-                        Err(e) => util::fail(&ctx, sink, RpcStatusCode::INTERNAL, &e),
+            match self_inner.meta_data.get_volume_by_id(src_vol_id).await {
+                Some(src_vol) => {
+                    let node_res = self_inner.meta_data.get_node_by_id(&src_vol.node_id).await;
+                    if let Some(node) = node_res {
+                        let client = MetaData::build_worker_client(&node);
+                        let create_res = client
+                            .worker_create_snapshot_async(&req)
+                            .map_err(|e| (RpcStatusCode::INTERNAL, anyhow::Error::new(e)))?
+                            .await
+                            .context(format!(
+                                "failed to create snapshot name={} on node ID={}",
+                                snap_name, src_vol.node_id,
+                            ));
+                        match create_res {
+                            Ok(r) => return Ok(r),
+                            Err(e) => return Err((RpcStatusCode::INTERNAL, e)),
+                        }
+                    } else {
+                        warn!("failed to find node ID={} from etcd", src_vol.node_id);
+                        // Should we panic here??????
```

Comment:
Should we panic here?

Reply:
Let's try fast fail now. If panic here, we should also set panic handling as abort in Cargo.toml

Reply:
I think there is a bug in the original code https://github.com/datenlord/datenlord/blob/master/csi/src/controller.rs#L806
The function won't send grpc reply if match this line. 

I will suggest send a failed grpc reply instead of panic as my implementation.